Introduce priority class in the resource quota

vikaschoudhary16 opened the pull request:
**What this PR does / why we need it**:
Implements https://github.com/kubernetes/community/pull/933
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #48648 

**Special notes for your reviewer**:
Test Cases are still to be covered. Opening this PR to make discussion convenient with code references.
Will update test cases only after design PR has got merged.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Ability to quota resources by priority
```
/kind feature
/priority important-soon
/sig scheduling
/sig node
/cc @resouer @derekwaynecarr @sjenning @bsalamat @timstclair @aveshagarwal @ravisantoshgudimetla

vikaschoudhary16 committed:
Introduce priority class in the resource quota

resouer commented:
Thanks a lot @vikaschoudhary16 for volunteering to take this! 

As we slacked, I will do first founds of review while eventually API folks need to do the final pass.

resouer commented on plugin/pkg/admission/resourcequota/controller.go:
```diff
@@ -367,6 +367,50 @@ func limitedByDefault(usage api.ResourceList, limitedResources []resourcequotaap
 
+func matchScope(priorityClass string, limitedResources []resourcequotaapi.LimitedResource) bool {
+	match := false
+	for _, limitedResource := range limitedResources {
```
Isn't this check only make sense for `resource: pods`?

derekwaynecarr replied:
there are other scopes besides these that we would need to handle.

is there a reason we cant reuse the pod scopes func so it handles BestEffort, NotBestEffort, Terminating, NotTerminating matches as well?  the admin could configure this to use any valid scope.

soltysh commented:
/unassign

bsalamat commented on pkg/quota/evaluator/core/pods.go:
```diff
@@ -327,6 +377,21 @@ func isTerminating(pod *api.Pod) bool {
+		return false
+	}
+	//TODO: match after creating a set of priority classes from scope
+	return strings.Contains(scope, pod.Spec.PriorityClassName)
```
This looks problematic to me. For example, if scope is "system-critical" and Pod priority class name is "critical", this function returns true, but IIUC, it should return false.

bsalamat reviewed (commented):
Some early comments. I only did a quick pass for now.

vikaschoudhary16 commented:
@bsalamat addressed your comments. Thanks!

I am more interested in knowing if there are any usecases handling of which is missing or is not as intended in this PR.
Thanks!

derekwaynecarr commented:
@vikaschoudhary16 is this ready for review?  can you get the tests passing and i will take a pass?

derekwaynecarr commented:
/retest

derekwaynecarr commented on pkg/quota/evaluator/core/persistent_volume_claims.go:
```diff
@@ -85,6 +85,11 @@ func (p *pvcEvaluator) Constraints(required []api.ResourceName, item runtime.Obj
 	return nil
 }
 
+func (p *pvcEvaluator) GetPriorityClass(item runtime.Object) (*string, error) {
```
its not clear to me yet why GetPriorityClass needs to be defined in each evaluator yet rather than just scoped to the pod evaluator.

derekwaynecarr commented on pkg/quota/evaluator/core/pods.go:
```diff
@@ -116,6 +118,23 @@ func (p *podEvaluator) Constraints(required []api.ResourceName, item runtime.Obj
 		return fmt.Errorf("Unexpected input object %v", item)
 	}
 
+	// Pod level resources are often set during admission control
```
why is this section included?

we do have a validation step in admission now.

derekwaynecarr commented on pkg/quota/evaluator/core/persistent_volume_claims.go:
```diff
@@ -119,8 +124,8 @@ func (p *pvcEvaluator) Handles(a admission.Attributes) bool {
 // Matches returns true if the evaluator matches the specified quota with the provided input item
-func (p *pvcEvaluator) Matches(resourceQuota *api.ResourceQuota, item runtime.Object) (bool, error) {
-	return generic.Matches(resourceQuota, item, p.MatchingResources, generic.MatchesNoScopeFunc)
+func (p *pvcEvaluator) Matches(resourceQuota *api.ResourceQuota, item runtime.Object, matchPriorityScope bool) (bool, error) {
```
also not obvious to me yet why the additional bool is needed on Matches func calls.

derekwaynecarr commented on pkg/quota/generic/evaluator.go:
```diff
@@ -69,6 +69,9 @@ type ListFuncByNamespace func(namespace string) ([]runtime.Object, error)
 // MatchesScopeFunc knows how to evaluate if an object matches a scope
 type MatchesScopeFunc func(scope api.ResourceQuotaScope, object runtime.Object) (bool, error)
 
+// MatchesPriorityClassScopeFunc knows how to evaluate if an object matches a scope
```
why do we need a new scope func type?

this is a scope like any other scope and should be handled by modifying the MatchesScopeFunc already.

derekwaynecarr commented on pkg/quota/generic/evaluator.go:
```diff
@@ -81,22 +84,39 @@ func MatchesNoScopeFunc(scope api.ResourceQuotaScope, object runtime.Object) (bo
-func Matches(resourceQuota *api.ResourceQuota, item runtime.Object, matchFunc MatchingResourceNamesFunc, scopeFunc MatchesScopeFunc) (bool, error) {
+func Matches(resourceQuota *api.ResourceQuota, item runtime.Object,
+	matchFunc MatchingResourceNamesFunc, scopeFunc MatchesScopeFunc,
+	priorityClassScopeFunc MatchesPriorityClassScopeFunc,
```
the signature for scopeFunc and priorityClassScope Func are equivalent

cant the pod Matches just pass a single scope func that knows how to evaluate all scope types?  in practice, pods are the only thing that use scopes now anyway.

vikaschoudhary16 replied:
We can merge both into one but then will have to pass `matchPriorityClassScopes` to the combined `scopeFunc`. If this bool val is true, then pod object must match one of the priority class scopes. To avoid this complexity and reducing `if-else`, thought of keeping separate funcs. I do realize that still there is new `if` block being introduced in [`Matches`](https://github.com/kubernetes/kubernetes/pull/57963/files#diff-df7430c6b773416a0aa76f1f627d1405R107).

Please suggest if you think otherwise.

derekwaynecarr commented on pkg/quota/generic/evaluator.go:
```diff
@@ -81,22 +84,39 @@ func MatchesNoScopeFunc(scope api.ResourceQuotaScope, object runtime.Object) (bo
+func Matches(resourceQuota *api.ResourceQuota, item runtime.Object,
+	matchFunc MatchingResourceNamesFunc, scopeFunc MatchesScopeFunc,
+	priorityClassScopeFunc MatchesPriorityClassScopeFunc,
+	matchPriorityClassScopes bool) (bool, error) {
```
why does this require a new bool?

vikaschoudhary16 replied:
explained in other comments.

derekwaynecarr commented on pkg/quota/interfaces.go:
```diff
@@ -50,13 +50,15 @@ type Evaluator interface {
 	Handles(operation admission.Attributes) bool
 	// Matches returns true if the specified quota matches the input item
-	Matches(resourceQuota *api.ResourceQuota, item runtime.Object) (bool, error)
+	Matches(resourceQuota *api.ResourceQuota, item runtime.Object, matchPriorityClassScopes bool) (bool, error)
```
i dont understand the reason for the bool argument.

the interface shouldnt need to change.

derekwaynecarr replied:
i would like to keep all notion of priority out of the evaluator interface.

derekwaynecarr commented on pkg/quota/interfaces.go:
```diff
@@ -50,13 +50,15 @@ type Evaluator interface {
 	// UsageStats calculates latest observed usage stats for all objects
 	UsageStats(options UsageStatsOptions) (UsageStats, error)
+	// GetPriorityClass returns PriorityClassName, if assigned any, to the object
+	GetPriorityClass(item runtime.Object) (*string, error)
```
i also dont understand why this is required?  this should be scoped to the pod evaluator only.

derekwaynecarr commented on plugin/pkg/admission/resourcequota/controller.go:
```diff
@@ -382,6 +426,17 @@ func (e *quotaEvaluator) checkRequest(quotas []api.ResourceQuota, a admission.At
 	inputObject := a.GetObject()
+	var priorityClass *string
+	var objectRequiresQuotaStrictly bool
+	priorityClass, err := evaluator.GetPriorityClass(inputObject)
```
scopes are not specific to priority per earlier comments.

we should be able to structure this by just passing the input object into a generic matches function that works for all scopes.

derekwaynecarr commented on plugin/pkg/admission/resourcequota/controller.go:
```diff
@@ -436,7 +491,11 @@ func (e *quotaEvaluator) checkRequest(quotas []api.ResourceQuota, a admission.At
 	if len(interestingQuotaIndexes) == 0 {
-		return quotas, nil
+		if objectRequiresQuotaStrictly {
+			return quotas, fmt.Errorf("No matching quota found with the configured MatchScopes")
```
i am not sure i understand the logic here.

i would expect that this would be handled similar to the hasNoCoveringQuota section above.

the error message could be improved to say "insufficient quota to consume: <resources> that match following scopes: <scopes>"

derekwaynecarr reviewed (changes requested):
i have a number of questions and requests for changes to how this is structured.  it also needs to handle all possible scope types (not just the priority ones).

vikaschoudhary16 commented:
@derekwaynecarr Thanks for taking a look and asking relevant questions. I will shortly reply/address your comments.

vikaschoudhary16 reviewed (commented):
hi @derekwaynecarr
Please find my explanations inline. Looking forward for further comments.

CaoShuFeng commented on pkg/quota/evaluator/core/pods.go:
```diff
@@ -337,6 +366,26 @@ func isTerminating(pod *api.Pod) bool {
+	if pod.Spec.PriorityClassName == "" {
+		return false
+	}
+	classes := strings.Split(scope, ":")[1:]
```
Why PriorityClassNames are also splitted by ":"?
Should they be splited by "," ?

derekwaynecarr replied:
agree ^

bsalamat replied:
+1 and do they start with ":"?

vikaschoudhary16 replied:
@bsalamat like this for example `PriorityClassNameIn:another-priorityclass-name:cluster-services`

We cannot use `,` as a delimiter between class names on cli because `,` is delimiter between scopes.
PTAL at [@derekwaynecarr  comment](https://github.com/kubernetes/kubernetes/pull/57963#discussion_r182766953)
>$ kubectl create quota myquota --hard=pods=10 --scopes=BestEffort,Terminating,PriorityClassNameIn:foo,bar

bsalamat replied:
Thanks. Do we have any other feature that gets its list like this (i.e., with `:` delimiters)? If not, we should probably use what other features use.

vikaschoudhary16 replied:
@bsalamat I searched through [Kubectl cli reference](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands), but could not find an existing command which expresses a nested list i.e similar to "list of priority classes within list of scopes".

some commands from expressing lists are there are the below ones:
> kubectl create quota my-quota --hard=cpu=1,memory=1G,pods=2,services=3,replicationcontrollers=2,resourcequotas=1,secrets=5,persistentvolumeclaims=10
> kubectl create clusterrole foo --verb=get,list,watch --resource=rs.extensions
> kubectl create clusterrole foo --verb=get,list,watch --resource=pods,pods/status

But none seems to be expressing nested list.
Wondering if you have any suggestion to make on top of your mind.

/cc @derekwaynecarr

vikaschoudhary16 replied:
ping @bsalamat @derekwaynecarr

aveshagarwal replied:
yeah dont know if there is any better way here that includes ',' to represent nested list.

vikaschoudhary16 replied:
@bsalamat How do we represent case like this, `scopes=PriorityClassNameIn=class1,class2,class3,BestEffort,Terminating` ?

bsalamat replied:
For the record, we chatted offline and decided to go with:
`scopes=PriorityClassNameIn=class1;class2;class3,BestEffort,Terminating`

CaoShuFeng commented on plugin/pkg/admission/resourcequota/apis/resourcequota/types.go:
```diff
@@ -54,4 +54,6 @@ type LimitedResource struct {
 	// with any storage class, the list would include
 	// ".storageclass.storage.k8s.io/requests.storage"
 	MatchContains []string
+	// +optional
```
Please add a comment for "MatchScopes".

CaoShuFeng commented on plugin/pkg/admission/resourcequota/controller.go:
```diff
@@ -382,6 +402,14 @@ func (e *quotaEvaluator) checkRequest(quotas []api.ResourceQuota, a admission.At
 	inputObject := a.GetObject()
 
+	// Check if object matches AdmissionConfiguration matchScopes
+	limitedScopes, err := getMatchedLimitedScopes(evaluator, inputObject, e.config.LimitedResources)
```
Why the MatchScopes and MatchContains are managed separately?
I think they can be managed together.

vikaschoudhary16 replied:
Since Scopes and resources are conceptually different and required different logic at evaluators as well, it is prefered to keep seperate APIs for each. Both may be managed together but that might make implementation complex.

CaoShuFeng replied:
OK :+1:

derekwaynecarr commented on pkg/quota/evaluator/core/pods.go:
```diff
@@ -279,6 +298,16 @@ func podMatchesScopeFunc(scope api.ResourceQuotaScope, object runtime.Object) (b
+	case api.ResourceQuotaScopePriorityClassNameNotExists:
+		return !doesPriorityClassExist(pod), nil
+	default:
+		if strings.Contains(string(scope), string(api.ResourceQuotaScopePriorityClassNameIn)) {
```
agree w/ @CaoShuFeng

derekwaynecarr commented on pkg/quota/evaluator/core/pods.go:
```diff
@@ -337,6 +366,26 @@ func isTerminating(pod *api.Pod) bool {
 }
 
+func doesPriorityClassExist(pod *api.Pod) bool {
+	if pod.Spec.PriorityClassName != "" {
```
simplify this to just

`return pod.Spec.PriorityClassName != ""`

derekwaynecarr commented on pkg/quota/interfaces.go:
```diff
@@ -51,6 +51,8 @@ type Evaluator interface {
 	Handles(operation admission.Attributes) bool
 	// Matches returns true if the specified quota matches the input item
 	Matches(resourceQuota *api.ResourceQuota, item runtime.Object) (bool, error)
+	// MatchingScopes takes the input specified list of scopes and pod object. Returns the set of scopes pod matches.
```
nit: MatchingScopes takes the input specified list of scopes and and returns the set of scopes that matches the input object.

derekwaynecarr reviewed (changes requested):
a few minor nits, but this is looking much better.

poke when handled.

vikaschoudhary16 commented:
@derekwaynecarr PTAL!!

derekwaynecarr commented on pkg/quota/evaluator/core/pods.go:
```diff
@@ -337,6 +366,23 @@ func isTerminating(pod *api.Pod) bool {
+	if pod.Spec.PriorityClassName == "" {
+		return false
+	}
+	classes := strings.Split((strings.Split(scope, ":")[1]), ",")
```
i am sorry for missing this earlier.

i am concerned that the `,` delimiter won't work when using the CLI.

for example:

```
$ kubectl create quota myquota --hard=pods=10 --scopes=BestEffort,Terminating,PriorityClassNameIn:foo,bar
```

this won't work.  we probably need to stick with a : delimiter.

derekwaynecarr reviewed (changes requested):
see the CLI comment, sorry :-(

vikaschoudhary16 commented:
/test pull-kubernetes-integration

vikaschoudhary16 commented:
ping @derekwaynecarr

derekwaynecarr commented:
/approve
/lgtm

derekwaynecarr commented:
for reference to api reviewers, see: https://github.com/kubernetes/community/pull/933

the usage of scopes was chosen as we lacked field selectors in user api.

vikaschoudhary16 commented on pkg/quota/evaluator/core/pods.go:
```diff
@@ -337,6 +366,26 @@ func isTerminating(pod *api.Pod) bool {
 	return false
 }
 
+func doesPriorityClassExist(pod *api.Pod) bool {
```
sure

vikaschoudhary16 commented:
/test pull-kubernetes-verify

aveshagarwal commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4555,6 +4555,7 @@ func ValidateResourceRequirements(requirements *core.ResourceRequirements, fldPa
 // validateResourceQuotaScopes ensures that each enumerated hard resource constraint is valid for set of scopes
 func validateResourceQuotaScopes(resourceQuotaSpec *core.ResourceQuotaSpec, fld *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
+	glog.V(4).Infof("scopes %v", resourceQuotaSpec.Scopes)
```
nit: do you really want to leave this glog msg?

vikaschoudhary16 replied:
I was thinking that this log message will not repeat again and again and will not spam the log file. OTOH, it might be useful in some misconfiguration cases.

bsalamat replied:
I am fine if you want to keep it, but please raise the verbosity levels to at least 6 and give it more descriptive text.

aveshagarwal commented on pkg/quota/evaluator/core/pods.go:
```diff
@@ -181,6 +183,47 @@ func (p *podEvaluator) MatchingResources(input []api.ResourceName) []api.Resourc
+
+// DifferenceScopes takes the input matched scopes which are limited by configuration and the matched quota scopes.
+// It returns the scopes which are in limited scopes but dont have a corresponding covering quota scope
+func (p *podEvaluator) DifferenceScopes(limitedScopes []api.ResourceQuotaScope, matchedQuotaScopes []api.ResourceQuotaScope) ([]api.ResourceQuotaScope, error) {
```
nit: how about UncoveredScopes or NonQuotaPriorityScopes or something else?

aveshagarwal commented on pkg/quota/generic/evaluator.go:
```diff
@@ -81,7 +81,8 @@ func MatchesNoScopeFunc(scope api.ResourceQuotaScope, object runtime.Object) (bo
 // Matches returns true if the quota matches the specified item.
-func Matches(resourceQuota *api.ResourceQuota, item runtime.Object, matchFunc MatchingResourceNamesFunc, scopeFunc MatchesScopeFunc) (bool, error) {
+func Matches(resourceQuota *api.ResourceQuota, item runtime.Object,
+	matchFunc MatchingResourceNamesFunc, scopeFunc MatchesScopeFunc) (bool, error) {
```
nit: is it intentional?

vikaschoudhary16 replied:
@aveshagarwal yes. for breaking very long line into two.

aveshagarwal commented on plugin/pkg/admission/resourcequota/apis/resourcequota/types.go:
```diff
@@ -54,4 +54,15 @@ type LimitedResource struct {
+	// quota system will ensure that there is a covering quota.  In the
+	// absence of a covering quota, the quota system will deny the request.
+	// For example, if an administrator wants to globally enforce that
+	// that a quota must exist to create a pod with "cluster-services" priorityclass
```
nit: that that

aveshagarwal commented on plugin/pkg/admission/resourcequota/controller.go:
```diff
@@ -403,13 +429,27 @@ func (e *quotaEvaluator) checkRequest(quotas []api.ResourceQuota, a admission.At
+
 		match, err := evaluator.Matches(&resourceQuota, inputObject)
 		if err != nil {
+			glog.V(5).Infof("err %v", err)
```
If its an error, why glgo.Infof?

aveshagarwal commented on plugin/pkg/admission/resourcequota/controller.go:
```diff
@@ -403,13 +429,27 @@ func (e *quotaEvaluator) checkRequest(quotas []api.ResourceQuota, a admission.At
+		for _, scope := range localRestrictedScopes {
+			restrictedScopes = append(restrictedScopes, scope)
+		}
+		glog.V(5).Infof("localRestrictedScopes %v", localRestrictedScopes)
```
Wonder if you intentionally put several glog infof msg or forgot to remove?

aveshagarwal commented:
@vikaschoudhary16 just some nits, otherwise lgtm (anyway this PR already has lgtm).

ravisantoshgudimetla commented:
/retest

bsalamat commented on plugin/pkg/admission/resourcequota/controller.go:
```diff
@@ -403,13 +429,27 @@ func (e *quotaEvaluator) checkRequest(quotas []api.ResourceQuota, a admission.At
 	// was limited by default.
 	restrictedResourcesSet := sets.String{}
+	restrictedScopes := []api.ResourceQuotaScope{}
+	glog.V(5).Infof("Going to iterate over quotas")
```
Please remove

bsalamat commented on plugin/pkg/admission/resourcequota/controller.go:
```diff
@@ -424,6 +464,7 @@ func (e *quotaEvaluator) checkRequest(quotas []api.ResourceQuota, a admission.At
 		interestingQuotaIndexes = append(interestingQuotaIndexes, i)
 		localRestrictedResourcesSet := quota.ToSet(restrictedResources)
 		restrictedResourcesSet.Insert(localRestrictedResourcesSet.List()...)
+
```
Please remove the blank line

vikaschoudhary16 reviewed (commented):
Addressed all the comments in latest commit update.

vikaschoudhary16 commented:
/test pull-kubernetes-e2e-gce

bsalamat commented on pkg/quota/evaluator/core/pods.go:
```diff
@@ -337,6 +387,23 @@ func isTerminating(pod *api.Pod) bool {
 }
 
+func hasPriorityClass(pod *api.Pod) bool {
+	return pod.Spec.PriorityClassName != ""
```
Please use: `len(pod.Spec.PriorityClassName) > 0` instead.

bsalamat commented on pkg/quota/evaluator/core/pods.go:
```diff
@@ -337,6 +387,23 @@ func isTerminating(pod *api.Pod) bool {
+}
+
+func isPriorityClassIn(pod *api.Pod, scope string) bool {
+	if pod.Spec.PriorityClassName == "" {
```
Please replace with: `if !hasPriorityClass(pod) {`

bsalamat commented on plugin/pkg/admission/resourcequota/controller.go:
```diff
@@ -403,10 +429,21 @@ func (e *quotaEvaluator) checkRequest(quotas []api.ResourceQuota, a admission.At
+
 		match, err := evaluator.Matches(&resourceQuota, inputObject)
 		if err != nil {
+			glog.Errorf("err %v", err)
```
Please give it a more descriptive text.

bsalamat commented on pkg/quota/evaluator/core/pods.go:
```diff
@@ -337,6 +387,23 @@ func isTerminating(pod *api.Pod) bool {
+		return false
+	}
+	classes := strings.Split(scope, "=")[1]
+	for _, class := range strings.Split(classes, ":") {
```
please replace the `:` with `;`. A semicolon is a better delimiter here IMO.

bsalamat reviewed (commented):
Thanks, @vikaschoudhary16! It looks good. I have just a few minor comments.

bsalamat reviewed (commented):
Thanks, @vikaschoudhary16! It looks good. I have just a few minor comments.

bsalamat reviewed (commented):
Thanks, @vikaschoudhary16! It looks good. I have just a few minor comments.

ravisantoshgudimetla commented:
xref: https://github.com/kubernetes/kubernetes/issues/57471

bsalamat reviewed (commented):
/lgtm

Thanks a lot, @vikaschoudhary16!

vikaschoudhary16 commented:
ping @derekwaynecarr @smarterclayton

vikaschoudhary16 commented:
@smarterclayton updated the log message. PTAL

vikaschoudhary16 commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu
/test pull-kubernetes-e2e-gce

vikaschoudhary16 commented:
/test pull-kubernetes-e2e-gce

derekwaynecarr commented on plugin/pkg/admission/resourcequota/apis/resourcequota/types.go:
```diff
@@ -54,4 +54,15 @@ type LimitedResource struct {
+	// the list would include
+	// "PriorityClassNameIn=cluster-services"
+	// +optional
+	MatchScopes []string `json:"matchScopes,omitempty"`
```
at this point, strongly coupling quota with priority is not the end of the world.

```
PrioritySelector struct {
  PriortySelectorTerms []PrioritySelectorTerm
}

ResourceQuotaSpec struct {
// scope quota to priority, resources must match pod names
Priority *PrioritySelector
... 
```

in the end, this may just be simpler for end users.

vikaschoudhary16 commented:
@derekwaynecarr @smarterclayton Updated the changes as per your inputs.PTAL!

vikaschoudhary16 commented:
/test pull-kubernetes-e2e-kops-aws

bsalamat commented on plugin/pkg/admission/resourcequota/admission_test.go:
```diff
@@ -1439,9 +1443,331 @@ func TestAdmitLimitedResourceWithQuotaThatDoesNotCover(t *testing.T) {
 }
+
+// TestAdmitLimitedScopeWithQuota verifies if a limited scope is configured the quota must cover the resource.
+func TestAdmitLimitedScopeWithCoverQuota(t *testing.T) {
```
Could you please add a test to limit a priority class globally and allow it only in one namespace and verify that admission works correctly?

vikaschoudhary16 replied:
Hi @bsalamat, following tests do this testing:
https://github.com/vikaschoudhary16/kubernetes/blob/f382229ccabd6b8b137778b674b8960a322e3718/plugin/pkg/admission/resourcequota/admission_test.go#L1705-#L1771

bsalamat replied:
The scenario that I described is not covered by those tests. The test at line 1755 is the closest one. The quota in that test has no priority class name though. It would be great if you could add a test very similar to the one at line 1705 (not 1755) and just change the pod's priority class to "another-priorityclass-name" and verify that there is no error. Thanks!

derekwaynecarr replied:
can we add a test case that validates the same resource "pods" being limited multiple times by varying scopes?

for example:
- "pods" with matching scope "Terminating"
- "pods" with matching scope "PriorityClass In Bar"

ensure that a pod that matches neither scope is allowed as expected, but a pod that matches either is?  it looks like each test tests a single LimitedResource but having multiple input LimitedResource that varies on scope would help.

vikaschoudhary16 replied:
@derekwaynecarr Added the following test cases:
```
1595:			description: "Two scopes,BestEffort and PriorityClassIN, in two LimitedResources. Neither matches pod. Pod allowed",
1624:			description: "Two scopes,BestEffort and PriorityClassIN, in two LimitedResources. Only BestEffort scope matches pod. Pod admit fails because covering quota is missing for BestEffort scope",
1653:			description: "Two scopes,BestEffort and PriorityClassIN, in two LimitedResources. Only PriorityClass scope matches pod. Pod admit fails because covering quota is missing for PriorityClass scope",
1682:			description: "Two scopes,BestEffort and PriorityClassIN, in two LimitedResources. Both the scopes matches pod. Pod admit fails because covering quota is missing for PriorityClass scope and BestEffort scope",
1711:			description: "Two scopes,BestEffort and PriorityClassIN, in two LimitedResources. Both the scopes matches pod. Quota available only for BestEffort scope. Pod admit fails because covering quota is missing for PriorityClass scope",
1753:			description: "Two scopes,BestEffort and PriorityClassIN, in two LimitedResources. Both the scopes matches pod. Quota available only for PriorityClass scope. Pod admit fails because covering quota is missing for BestEffort scope",
1795:			description: "Two scopes,BestEffort and PriorityClassIN, in two LimitedResources. Both the scopes matches pod. Quota available only for both the scopes. Pod admit success. No Error",
```

vikaschoudhary16 replied:
@bsalamat Added following test case as per your suggestion:
```
1995:                         description: "From the above test case, just changing pod priority from cluster-services to another-priorityclass-name. expecting no error"
```

bsalamat replied:
Thanks! Looks good.

derekwaynecarr commented on pkg/apis/core/types.go:
```diff
@@ -4108,8 +4110,46 @@ type ResourceQuotaSpec struct {
 	Scopes []ResourceQuotaScope
+	// Priority at which this quota object is scoped
+	// +optional
+	Priority *ScopeSelector
```
should this be `ScopeSelector` rather than `Priority`

derekwaynecarr replied:
ok, i think this should be `scopeSelector` since it has backwards compatibility with prior scopes via exists operator.

derekwaynecarr commented on pkg/quota/generic/evaluator.go:
```diff
@@ -90,12 +92,21 @@ func Matches(resourceQuota *api.ResourceQuota, item runtime.Object, matchFunc Ma
 	matchScope := true
 	for _, scope := range resourceQuota.Spec.Scopes {
-		innerMatch, err := scopeFunc(scope, item)
+		innerMatch, err := scopeFunc(api.ScopedResourceSelectorRequirement{ScopeName: scope}, item)
```
so this conversion treats the operator as optional?

derekwaynecarr commented on pkg/quota/resources.go:
```diff
@@ -252,6 +252,15 @@ func ToSet(resourceNames []api.ResourceName) sets.String {
 }
 
+// ScopesToSet takes a list of scopes and converts to a string set
+func ScopesToSet(scopes []api.ResourceQuotaScope) sets.String {
```
it doesn't look like anyone calls this?

derekwaynecarr commented on plugin/pkg/admission/resourcequota/controller.go:
```diff
@@ -516,6 +560,21 @@ func (e *quotaEvaluator) checkRequest(quotas []api.ResourceQuota, a admission.At
+	for _, scope := range quota.Spec.Scopes {
+		selectors = append(selectors, api.ScopedResourceSelectorRequirement{
+			ScopeName: scope,
+			Operator:  api.ScopeSelectorOpExists})
```
ah ok. operator defaults to existence.  so this provides backwards compatibility with old syntax.

derekwaynecarr reviewed (changes requested):
this is a change for the better, we need validation code similar to ValidateNodeSelectorTerm.  Also think the field should be renamed `scopeSelector`.

derekwaynecarr commented on pkg/apis/core/types.go:
```diff
@@ -4097,6 +4097,8 @@ const (
 	// Match all pod objects that do not have best effort quality of service
 	ResourceQuotaScopeNotBestEffort ResourceQuotaScope = "NotBestEffort"
+	// Match all pod objects that have priority class mentioned
+	ResourceQuotaScopePriorityClass ResourceQuotaScope = "PriorityClass"
```
update `IsResourceQuotaScopeValidForResource` to say that priority is valid for both pod object count and pod compute quota resources.

derekwaynecarr replied:
noting to myself this is done.

derekwaynecarr commented on pkg/quota/evaluator/core/pods.go:
```diff
@@ -265,12 +301,12 @@ func toInternalPodOrError(obj runtime.Object) (*api.Pod, error) {
 		return false, err
 	}
-	switch scope {
+	switch selector.ScopeName {
```
if the new syntax matches an exists operation on the old syntax, the other cases probably need to be updated at a future time.  if you use the new syntax, we lack some protection with what invalidScopePairs did in validation to find conflicting scopes.  if validation ensured that exists operator was only valid for the other types, i think this would keep the code logically consistent.  WDYT?

derekwaynecarr replied:
where other types == other scopes (duration and qos related)

vikaschoudhary16 replied:
I think you meant "if the validation ensured that for other types only exists operator was valid, ..." ?

vikaschoudhary16 replied:
updated `validateScopeSelector` in validation.go to ensure that if "other scopes" are being expressed in quota object using scope selector, any existing validation logic is not missed.

derekwaynecarr replied:
correct, this looks good now.

derekwaynecarr commented on pkg/quota/generic/evaluator.go:
```diff
@@ -76,29 +76,57 @@ type UsageFunc func(object runtime.Object) (api.ResourceList, error)
 		}
 		matchScope = matchScope && innerMatch
 	}
+	//if resourceQuota.Spec.Priority != nil {
```
vestigial code?

vikaschoudhary16 replied:
yep, removed.

derekwaynecarr commented on pkg/quota/interfaces.go:
```diff
@@ -32,6 +32,7 @@ type UsageStatsOptions struct {
 	Scopes []api.ResourceQuotaScope
 	// Resources are the set of resources to include in the measurement
 	Resources []api.ResourceName
+	Priority  *api.ScopeSelector
```
should be `ScopeSelector`

derekwaynecarr reviewed (changes requested):
few more requests.

do you think we need an e2e test added specific for priority here:

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/scheduling/resource_quota.go

that covers pod with priority?

vikaschoudhary16 commented:
@derekwaynecarr 
> do you think we need an e2e test added specific for priority, that covers pod with priority?

100% we should add. In the same PR or in a follow-on PR?

vikaschoudhary16 commented:
/test pull-kubernetes-integration

vikaschoudhary16 commented:
/test pull-kubernetes-e2e-gce

derekwaynecarr commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4735,6 +4735,65 @@ func validateResourceQuotaScopes(resourceQuotaSpec *core.ResourceQuotaSpec, fld
+	}
+	fldPath := fld.Child("scopeSelector")
+	scopeSet := sets.NewString()
+	for _, req := range resourceQuotaSpec.ScopeSelector.MatchExpressions {
```
nit: but the content in this for loop is easier to follow if we had a `validateScopedResourceSelectorRequirement`

derekwaynecarr commented on plugin/pkg/admission/resourcequota/admission_test.go:
```diff
@@ -58,12 +59,15 @@ func getResourceRequirements(requests, limits api.ResourceList) api.ResourceRequ
 }
 
-func validPod(name string, numContainers int, resources api.ResourceRequirements) *api.Pod {
+func validPod(name string, numContainers int, resources api.ResourceRequirements, priorityClass string) *api.Pod {
```
this is a style thing, but looking back would have preferred we could have chained this call.

`validPod(foo, 3).WithResources(xyz).WithPriority(foo)`

derekwaynecarr replied:
or `validPod(foo, 3, resources).WithPriority(foo)`

derekwaynecarr reviewed (approved):
i am fine with the new e2e added in a follow-on since unit test coverage looks ok in this first pr and existing e2es ensure no regression.

my issues are all nits, this is good for me.

/approve
/lgtm

vikaschoudhary16 commented:
@derekwaynecarr Addressed the nits!!

dims commented:
/test pull-kubernetes-kubemark-e2e-gce-big

derekwaynecarr commented:
/approve
/lgtm

vikaschoudhary16 commented:
/test pull-kubernetes-e2e-gce

sjenning commented:
/lgtm

smarterclayton commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -11508,6 +11508,18 @@ func TestValidateResourceQuota(t *testing.T) {
+				},
+			},
+		},
+	}
```
Probably needs more validation tests but I'd be ok with those being later.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4710,8 +4712,47 @@ type ResourceQuotaSpec struct {
 	// If not specified, the quota matches all objects.
 	// +optional
 	Scopes []ResourceQuotaScope `json:"scopes,omitempty" protobuf:"bytes,2,rep,name=scopes,casttype=ResourceQuotaScope"`
+	// ScopeSelector is also a collection of filters like Scopes that must match each object tracked by a quota
```
This should be lower cased (publicly accessible API docs should use the field names an end user would see).  `scopeSelector ... scopes ...`.

As a follow up, I'd like to see a much better godoc here that describes how a user would use this.  Can be done after this merges.

smarterclayton replied:
You need to define whether this is `OR` with Scopes or `AND` with Scopes.

smarterclayton reviewed (changes requested):
Need to update the godoc with casing.

vikaschoudhary16 commented:
/test pull-kubernetes-e2e-gce-100-performance

sjenning commented:
/lgtm

smarterclayton commented:
Oops.  One more thing - put the API fields behind a feature flag (since I assumed this isn't going straight to GA) to prevent them being saved if the flag is off.  Same as initializers.  Mark the field as alpha.

smarterclayton commented:
Alpha for 1.11.  Feature gate looks good.  API looks good.  Assuming passing tests:

/approve
/lgtm

zparnold commented:
Hello there! @vikaschoudhary16  I'm Zach Arnold working on Docs for the 1.11 release. This PR was identified as one needing some documentation in the https://github.com/kubernetes/website repo around your contributions (thanks by the way!) When you have some time, could you please modify/add/remove the relevant content that needs changing in our documentation repo? Thanks! Please let me or my colleague Misty know (@zparnold/@Misty on K8s Slack) if you need any assistance with the documentation.
